go/vt/vtgate/planbuilder/operators/projection_pushing.go

```diff
@@ -0,0 +1,414 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operators
+
+import (
+	"slices"
+	"strconv"
+
+	"vitess.io/vitess/go/slice"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
+	"vitess.io/vitess/go/vt/vtgate/semantics"
+)
+
+type (
+	projector struct {
+		columns               []*ProjExpr
+		columnAliases         []string
+		explicitColumnAliases bool
+		tableName             sqlparser.TableName
+	}
+)
+
+// add introduces a new projection with the specified alias to the projector.
+func (p *projector) add(pe *ProjExpr, alias string) {
+	p.columns = append(p.columns, pe)
+	if alias != "" && slices.Index(p.columnAliases, alias) > -1 {
+		panic("alias already used")
+	}
+	p.columnAliases = append(p.columnAliases, alias)
+}
+
+// get finds or adds an expression in the projector, returning its SQL representation with the appropriate alias
+func (p *projector) get(ctx *plancontext.PlanningContext, expr sqlparser.Expr) sqlparser.Expr {
+	for _, column := range p.columns {
+		if ctx.SemTable.EqualsExprWithDeps(expr, column.EvalExpr) {
+			alias := p.claimUnusedAlias(column.Original)
+			out := sqlparser.NewColName(alias)
+			out.Qualifier = p.tableName
+
+			ctx.SemTable.CopySemanticInfo(expr, out)
+			return out
+		}
+	}
+
+	// we could not find the expression, so we add it
+	alias := sqlparser.UnescapedString(expr)
+	pe := newProjExpr(sqlparser.NewAliasedExpr(expr, alias))
+	p.columns = append(p.columns, pe)
+	p.columnAliases = append(p.columnAliases, alias)
+
+	out := sqlparser.NewColName(alias)
+	out.Qualifier = p.tableName
+
+	ctx.SemTable.CopySemanticInfo(expr, out)
+
+	return out
+}
+
+// claimUnusedAlias generates a unique alias based on the provided expression, ensuring no duplication in the projector
+func (p *projector) claimUnusedAlias(ae *sqlparser.AliasedExpr) string {
+	bare := ae.ColumnName()
+	alias := bare
+	var i int64 = 0
+
+	for slices.Index(p.columnAliases, alias) > -1 {
+		alias = bare + strconv.FormatInt(i, 10)
+		i++
+	}
```

Comment:
being picky about coding style here, but you could also inline the definition and incrementation of `i` in the for loop as follows:
```suggestion
	for i := int64(0); slices.Index(p.columnAliases, alias) > -1; i++ {
		alias = bare + strconv.FormatInt(i, 10)
	}
```